src/signature.c

```diff
@@ -64,14 +65,36 @@ static char *crl = NULL;
  * returns: true if can initialize and validate certificates, otherwise false */
 bool initialize_signature(void)
 {
+	int ret = -1;
+	time_t mod_sec = 0;
+	struct tm * alttime;
+	struct stat statt;
 	ERR_load_crypto_strings();
 	ERR_load_PKCS7_strings();
 	EVP_add_digest(EVP_sha256());
 
 	if (!get_pubkey()) {
 		goto fail;
 	}
-	if (!validate_certificate()) {
+
+	ret = validate_certificate();
+	if (ret) {
+		if (ret == X509_V_ERR_CERT_NOT_YET_VALID) {
+			/* If we can retrieve an approx. good system time, report out to user */
+			if (stat("/usr/lib/os-release", &statt) != -1) {
+				mod_sec = statt.st_mtim.tv_sec;
+				char timebuf[30];
+				alttime = localtime(&mod_sec);
+				strftime(timebuf, sizeof(timebuf), "%F", alttime);
+				printf("Failure: Certificate not yet valid. System clock should be at least %s\n",
+						timebuf);
+			} else {
+				printf("Failure: Certificate not yet valid, check system time\n");
+			}
+		} else {
+			printf("Failed to verify certificate: %s\n", X509_verify_cert_error_string(ret));
```

Comment:
In the interest of making error strings discoverable, it would be nice for this `printf` to live on line 82. So the code would look like:

```
if (ret) {
	printf("Failed to verify certificate: %s\n", X509_verify_cert_error_string(ret));
	if (ret == X509_V_ERR_CERT_NOT_YET_VALID) {
		/* If we can retrieve an approx. good system time, report out to user */
		if (stat("/usr/lib/os-release", &statt) != -1) {
			mod_sec = statt.st_mtim.tv_sec;
			char timebuf[30];
			alttime = localtime(&mod_sec);
			strftime(timebuf, sizeof(timebuf), "%F", alttime);
			printf("System clock should be at least %s\n", timebuf);
		}
	}

	goto fail;
}
```